Schemas: Add basic events

d00rman opened the pull request:
This PR proposes the schemas for the basic set of events emitted by MediaWiki:
- article edit (and create)
- article delete
- article undelete
- article rename
- revision visibility change

Bug: [T116247](https://phabricator.wikimedia.org/T116247)

Commit:
Schemas: Add basic events

This commit proposes the schemas for the basic set of events emitted by
MediaWiki:

- article edit (and create)
- article delete
- article undelete
- article rename
- revision visibility change

d00rman commented:
Note: there is currently no separate article create event on the account of the fact that article edit events will fire for that one too. Nevertheless, it might be worth adding that separate event for completeness. Thoughts?

Pchelolo commented on schemas/mw-article-rename.js:
```diff
@@ -0,0 +1,48 @@
+    "required": [
+        "uri", "request_id", "dt", "domain", "title", "old_title", "page_id"
+    ]
+}
```
`old_revision` and `new_revision` fields might be helpful too, because when article is moved, a new revision of an old title with a redirect is created, so it wouldn't be obvious, which revision is a base for the move.

d00rman replied:
Good point. Will add.

ottomata commented on schemas/mw-article-delete.js:
```diff
@@ -0,0 +1,49 @@
+            "type": "string",
+            "description": "the title of the article"
+        },
+        "pageid": {
```
In Analytics world, we have been calling this `page_id`.  You use this for Edit too.

d00rman replied:
Ups, forgot the underscore.

ottomata commented on schemas/mw-article-delete.js:
```diff
@@ -0,0 +1,49 @@
+            "description": "the domain the event pertains to"
+        },
+        // event-specific fields
+        "title": {
```
A little redundant, but can we call this `page_title`?  Its called this in the (newish) canonical `pageviews_hourly` table:
https://wikitech.wikimedia.org/wiki/Analytics/Data/Pageview_hourly

d00rman replied:
I have no opinion on that, other than, as you point out, it's redundant. Would changing the name in pageviews be a big deal?

ottomata replied:
Ha, yes.

ottomata replied:
Event if it is redundant, we like longer names so the fields make sense and are easy to understand outside of a particular context.  This is relevant here since JSON Schema has a special `title` field that we are using.

gwicke replied:
JSON schema and an event instance are separate namespaces, so there is no risk of confusion. I personally think that `title` is a fine name for this attribute.

ottomata commented on schemas/mw-article-edit.js:
```diff
@@ -0,0 +1,63 @@
+            "description": "the domain the event pertains to"
+        },
+        // event-specific fields
+        "title": {
```
`page_title`?

ottomata commented on schemas/mw-article-edit.js:
```diff
@@ -0,0 +1,63 @@
+            "minimum": 1,
+            "description": "the revision ID that the edit is based on"
+        },
+        "save_time": {
```
Just an idea:

In a past life, when I was in charge of MySQL schema design, we adopted a convention that meaningful timestamps should end in `_at`.  So this could be `saved_at`.  If `save_time` is already a Mediawiki thing used elsewhere, please disregard and stick with `save_time`.

halfak replied:
It's called "timestamp" anywhere I can think of.  I don't want to have to write documentation that says "When you see 'save_time' or 'saved_at' in the event, that corresponds to 'rev_timestamp' in the database and '<timestamp>' in the XML dumps."  IMO, we shouldn't be in the business of fixing MediaWiki's terminology unless we are willing to do it _everywhere_.  Otherwise, we're just introducing yet another inconsistency.

Ironholds replied:
Even EventLogging refers to it as 'timestamp'. Timestamp is probably fine, yeah.

halfak commented:
Generally, we refer to objects in MediaWiki as pages.  "Article" is a word reserved for the main namespace on Wikipedias.  All instances of "article" should be replaced with "page".

halfak commented:
Also, it seems that we're inventing a lot of new terminology.  MediaWiki refers to the previous revision of the page via a "parent_id".  The "parent" revision is not necessarily the version that an edit is based on -- especially in the case of a revert.  

Further, "save_time" or "saved_at" are strange language for anyone familiar with MediaWiki data.  Can an event just have a "timestamp"?

halfak commented on schemas/mw-article-delete.js:
```diff
@@ -0,0 +1,54 @@
+            "minimum": 1,
+            "description": "the page ID of the deleted article"
+        },
+        "revision": {
```
A revision is created when an page is deleted?

Ironholds replied:
You could look at, say, the log_id for the deletion.

halfak replied:
+1 for that!

d00rman replied:
You are right, I'll remove the revision id. For the log id, I'm not sure how we could get it, so I'd propose to leave it out for now and add it later in case we can reliably set it.

d00rman replied:
Duh, the hook has the associated log entry :P

gwicke replied:
How will this log_id be used? Would it be more useful to directly include some or all of the corresponding log entry?

halfak commented on schemas/mw-article-edit.js:
```diff
@@ -0,0 +1,68 @@
+    "title": "MediaWiki Article Edit",
+    "description": "Represents a MW Article Edit event",
+    "type": "object",
+    "properties": {
```
Missing user info.

halfak replied:
Here are the fields I'd add:  user_id, user_text, byte_length/char_length, sha1, minor

d00rman replied:
I'm quite ambivalent towards the user info. In the case of anon edits, it will contain the user's IP and we wouldn't want to expose that. On the other hand, having user info in these events can come in quite handy.

Ironholds replied:
Would the bot status be worth adding?

halfak replied:
> In the case of anon edits, it will contain the user's IP 

Yes and so does the public interface.  This is part of standard MediaWiki behavior. 

> having user info in these events can come in quite handy.

It is essential -- just just handy.

halfak commented on schemas/mw-article-undelete.js:
```diff
@@ -0,0 +1,63 @@
+{
+    "title": "MediaWiki Article Undelete",
```
MediaWiki refers to this as a 'restore'.

d00rman commented:
> Also, it seems that we're inventing a lot of new terminology. MediaWiki refers to the previous revision of the page via a "parent_id". The "parent" revision is not necessarily the version that an edit is based on -- especially in the case of a revert.

Are you referring to the `base_revision` ? That's the exact point of it - to denote the the revision that the edit is based on (which, as you point out, may well not be the latest).

> Further, "save_time" or "saved_at" are strange language for anyone familiar with MediaWiki data. Can an event just have a "timestamp"?

In the context of events, I find //timestamp// to be ambiguous. Each event has a `dt` field, which is the time stamp of the event itself. Here, `saved_(time|at)` refers to the time stamp of when the user clicked. Granted, these two will differ only slightly in most cases, but it's an important distinction.

halfak commented on schemas/mw-article-rename.js:
```diff
@@ -0,0 +1,63 @@
+{
+    "title": "MediaWiki Article Rename",
```
MediaWiki refers to this as a 'move'

halfak commented on schemas/mw-article-undelete.js:
```diff
@@ -0,0 +1,63 @@
+            "type": "string",
+            "description": "the title of the article"
+        },
+        "page_id": {
```
We should explicitly call this 'new_page_id'.

d00rman replied:
AFAIK, yes. Restoring a page creates a revision that says *restored page blah*

halfak commented on schemas/mw-article-undelete.js:
```diff
@@ -0,0 +1,63 @@
+            "type": "integer",
+            "description": "the namespace ID the page belongs to"
+        },
+        "revision": {
```
'restore' events create a new revision?

d00rman replied:
AFAIK, yes. Restoring a page creates a revision that says _restored page blah_

halfak replied:
I just checked and I can't find any evidence of this on recent page restorations.

ottomata commented:
> Further, "save_time" or "saved_at" are strange language for anyone familiar with MediaWiki data. Can an event just have a "timestamp"?

@halfak, that's fair, although just `timestamp` isn't good for a couple of reasons:
- it is often a keyword, so not using it might avoid issues down the line
- The meaning of the field should be imparted in the name.  `timestamp` is too ambiguous.  It'd be like calling a string typed field `string`.

But, on the other hand, we should be as consistent as possible with existing Mediawiki fields.  There will be cases where we will choose new names for something Mediawiki uses, because Mediawiki names may be especially poorly chosen.  But, in the cases where they are at lease okay names, we should use them for consistency's sake.

Is there a better and more consistent name for `save_time`?

Ironholds commented:
On the based_on or timestamp; sure, they might not be the perfect names. But unless we have very good reasons to change them, we should keep them as they are - and when we do have to change them, do so minimally.

So, parent_id is what mediawiki calls it. Let's call it parent_id here. Timestamp is what mediawiki calls it; if we have issues with that, maybe we look at <code>dt</code> or <code>ts</code>, which is what we use in other places where timestamp is bad mojo.

halfak commented:
> Are you referring to the base_revision ? That's the exact point of it - to denote the the revision that the edit is based on (which, as you point out, may well not be the latest).

The base_revision is unknown to the MediaWiki software.  The parent_revision is known. 

> In the context of events, I find //timestamp// to be ambiguous. Each event has a dt field, which is the time stamp of the event itself. Here, saved_(time|at) refers to the time stamp of when the user clicked. Granted, these two will differ only slightly in most cases, but it's an important distinction.

That's fine that you find it to be ambiguous.  It's a standard term across all other datasets.  Consistency first!  Don't rewrite the naming scheme unless you are willing to fix _all of it_.

Ironholds commented on schemas/mw-article-delete.js:
```diff
@@ -0,0 +1,54 @@
+        },
+        "summary": {
+            "type": "string",
+            "description": "the summary comment left by the user"
```
You want log::log_comment here (same as the issue above re the revision id)

d00rman replied:
This is supposed to reflect [the `reason` field](https://github.com/wikimedia/mediawiki/blob/2cb813ff13f1b005fbb106b40fb9262430a528e1/docs/hooks.txt#L587) from the `ArticleDeleteComplete` hook.

Ironholds replied:
Gotcha - yeah, log_comment in a MW database sense.

Ironholds commented on schemas/mw-article-rename.js:
```diff
@@ -0,0 +1,63 @@
+        "page_id": {
+            "type": "integer",
+            "minimum": 1,
+            "description": "the page ID of the renamed article"
```
'moved'

d00rman replied:
yup

Ironholds commented on schemas/mw-article-rename.js:
```diff
@@ -0,0 +1,63 @@
+            "description": "the domain the event pertains to"
+        },
+        // event-specific fields
+        "title": {
```
Suggest new_title to contrast with old_title (and correspond with comments around new/old pageIDs below)

d00rman replied:
Makes sense.

Ironholds commented on schemas/mw-revision-visibilityset.js:
```diff
@@ -0,0 +1,57 @@
+            "minimum": 1,
+            "description": "the revision ID the visibility of which is being changed"
+        },
+        "sha1_hidden": {
```
Does the SHA1 _get_ hidden? I've seen the text, user and edit summary vanish - along with the whole revision - but not the hash. Mind you, I'm not sure how hash visibility would be apparent from the user side.

d00rman replied:
According to the [MW code](https://github.com/wikimedia/mediawiki/blob/2cb813ff13f1b005fbb106b40fb9262430a528e1/includes/api/ApiQueryRevisionsBase.php#L209-L221), this gets set automatically if the text is deleted.

Ironholds replied:
So, it automatically hides the SHA if the text is hidden. Ergo, the SHA being hidden is a proxy for text being hidden. And we're already tracking text being hidden. So...what's the use case for tracking both?

gwicke replied:
It might be a bit more compact / efficient to use an optional, nested `hidden` object instead. Example:

``` javascript
{
   ...,
   hidden: {
     text: true
   }
}
```

or, using an array:

``` javascript
{
   ...,
   hidden: ['text']
}
```

Advantages:
- Fast path can be a check for the presence of the 'hidden' object.
- Keys to hide are directly supplied in iterable fashion, without a need to hardcode a list in the consumer. This can make client code more compact (a loop), and lets us possibly extend suppression to other fields in the future.

gwicke commented:
I agree with @d00rman and @ottomata on 'timestamp' being not specific enough. Events will have multiple timestamps, and the EventBus will contain events from MediaWiki as well as other services, so lets use names that are clear in that context. For example, we could call it `rev_timestamp` or `mw_rev_timestamp`.

gwicke commented on schemas/mw-article-delete.js:
```diff
@@ -0,0 +1,54 @@
+            "pattern": "^[a-fA-F0-9]{8}(-[a-fA-F0-9]{4}){3}-[a-fA-F0-9]{12}$",
+            "description": "the unique ID of this event; added by the event bus"
+        },
+        "dt": {
```
Perhaps call this `event_ts`, to make it clear that it corresponds to `event_id`? It's more important to capture the semantics in the name. The type (unix timestamp vs. datetime) is already captured in the type, and will be obvious to anybody looking at actual data.

d00rman replied:
Makes sense. Will do.

gwicke replied:
We could also consider moving to a nested `event` object, as brought up by @ottomata and sketched in https://phabricator.wikimedia.org/T116247#1757410. This would simplify schema reuse and generic processing somewhat, and provide a logical grouping.

Commit:
Basic schemas: switch to YAML

d00rman commented:
I've made the following changes:
- moved global event fields into an `event` object
- added `user_id` and `user_text` to most events
- renamed `s/article/page/` in event names (and file names)
- `s/undelete/restore/` && `s/rename/move/`
- removed `revision` from the _mw-page-restore_ event
- there's now a `hidden` object in _mw-revision-visibilityset_ that holds info about all hidden fields

Also, I've switched the schema format to YAML to increase readability.

Pchelolo commented on schemas/mw-page-delete.yaml:
```diff
@@ -0,0 +1,54 @@
+required:
+  - event
+  - title
+  - revision
```
There's no `revision` property in the event.

Pchelolo replied:
Also, may be worths adding a `namespace` property?

d00rman replied:
Since this is the first version of the schemas, I'd really like to nail down the minimalistic set on which we can all agree but still allow us to continue our work on change propagation. In that light, I don't see _namespace_ to be an essential piece of information when deleting pages.

Pchelolo commented on schemas/mw-page-move.yaml:
```diff
@@ -0,0 +1,66 @@
+  - event
+  - new_title
+  - old_title
+  - page_id
```
It's possible to move the page across namespaces, so it could worth adding `old_namespace` and `new_namespace`

d00rman replied:
Idem. For the time being, we mostly care about whether a "page" is in the `NS_FILE` namespace or not. AFAIK, a page cannot be moved to/fro that one.

Pchelolo replied:
@d00rman true, we can't, I've just checked. 
If we would be able to add more fields later, I totally agree on keeping things simple at the beginning.

ottomata commented on schemas/mw-page-edit.yaml:
```diff
@@ -0,0 +1,70 @@
+    type: integer
+    minimum: 1
+    description: the revision ID that the edit is based on
+  save_time:
```
On more though, `save_time` is a bad name.  To me, it indicates the amount of time it took to save, not a timestamp.  `save_dt`?

d00rman replied:
Done!

Commit:
Minor: Page Edit: s/save_time/save_dt/

Commit:
Schemas: Load the schemas on start-up

JanZerebecki commented on schemas/mw-page-delete.yaml:
```diff
@@ -0,0 +1,53 @@
+title: MediaWiki Page Delete
```
After an event of this type, is it ok to still have the historic revisions of this page published?

d00rman replied:
AFAIK, that depends on user rights, so they'd need to be logged in. So I'd say _no, historic records for such a page should not be publicly available_.

Pchelolo replied:
More info on this topic is available here: https://en.wikipedia.org/wiki/Wikipedia:Viewing_deleted_content

gwicke replied:
We'll need to filter for public consumption.

ottomata commented on schemas/mw-page-delete.yaml:
```diff
@@ -0,0 +1,53 @@
+        type: string
+        pattern: '^[a-fA-F0-9]{8}(-[a-fA-F0-9]{4}){3}-[a-fA-F0-9]{12}$'
+        description: the unique ID of this event; should match the ts field
+      ts:
```
`dt`, right?

d00rman replied:
Haha, bike-shed forever. Sure, why not.

gwicke replied:
This is the hungarian notation vs. semantics debate ;)

ottomata commented on schemas/mw-revision-visibilityset.yaml:
```diff
@@ -0,0 +1,62 @@
+title: MediaWiki Revision Visibility Set
```
Maybe this is already a MW name, but if it isn't 'Set' here could be confusing.  To me, this title indicates that this record is for a set of 'revision visibilities' (whatever that might be).  Does MW call this a 'set event'?  If so, ignore me.  If not, maybe 'MediaWiki Revision Visibility Change'?

d00rman replied:
That's the name of the Hook emitted by MW.

halfak replied:
Docs around the log event call it "Change visibility".  See https://www.mediawiki.org/wiki/Manual:Log_actions.

ottomata commented on schemas/mw-page-delete.yaml:
```diff
@@ -0,0 +1,53 @@
+type: object
+properties:
+  # global event fields
+  event:
```
I think `event` is a misleading name for this meta data.  Can we call this `meta`?

d00rman replied:
Don't you find `meta` obscure? How about `event_meta` ?

gwicke replied:
Another option might be `head`, after the html element. I was also looking at http://stateless.co/hal_specification.html to see if we can steal any ideas there (esp. `_links`), but our head / meta / envelope info is a mix of references and data. We could shoehorn that data into URLs, but it doesn't feel super natural.

Commit:
Schemas: s/ts/dt/ and s/event/meta/

d00rman commented:
@ottomata did `s/ts/dt/` and `s/event/meta/`

milimetric commented:
Wanted to chime in on two small things:
- I think `request_id` should be renamed to x_request_id or http_request_id.  As it is, the fact that it comes from X-Request-Id is lost in the translation from that header to the event.  And it's confusing to people who understand "request" as a generic concept.
- I also think that `id` should be thought about carefully.  If a client is allowed to send it, we have to make sure all clients create compatible ids, and validating that would take special code (to make sure a valid timestamp can be pulled out of the id, etc.).  Also if a client is allowed to send it, all clients need to be checked to make sure they can't send duplicates, because that's not possible to validate either.  If, instead, the server generated that id, as it does in Event Logging, we could tightly control the logic of the uuid creation and ensure that the only duplicates are true duplicates.  This makes it easier to analyze the data later on, or to deal with de-duplication for non-analysis purposes.

I've not fully thought those two things through, just saying we need to.

d00rman commented:
> - I think request_id should be renamed to x_request_id or http_request_id. As it is, the fact that it comes from X-Request-Id is lost in the translation from that header to the event. And it's confusing to people who understand "request" as a generic concept.

The `X-` in the `X-Request-Id` header is only there because it's an HTTP convention to prefix non-standard headers with it, so from that perspective `request_id` is perfectly fine for me. Its purpose it to be able to track a request through the WMF architecture. Which concept are you referring to with _"request" as a generic concept_?

> - I also think that id should be thought about carefully. If a client is allowed to send it, we have to make sure all clients create compatible ids, and validating that would take special code (to make sure a valid timestamp can be pulled out of the id, etc.). Also if a client is allowed to send it, all clients need to be checked to make sure they can't send duplicates, because that's not possible to validate either. If, instead, the server generated that id, as it does in Event Logging, we could tightly control the logic of the uuid creation and ensure that the only duplicates are true duplicates. This makes it easier to analyze the data later on, or to deal with de-duplication for non-analysis purposes.

You are right that having these two fields can pose certain problems. Perhaps a good approach would be for the REST service to check them and make sure they point to the same moment in time. Additionally, we might allow users to specify only one of them, with the REST service filling in the other one.

milimetric commented:
> Which concept are you referring to with "request" as a generic concept?

web request, REST service request, client-side request, server-side request, requirements request, request for comment, request to go to the bathroom :)

"webrequest" is fine, just not simply "request".

ottomata commented on schemas/mw-page-delete.yaml:
```diff
@@ -0,0 +1,53 @@
+        description: the domain the event pertains to
+    required:
+      - uri
+      - request_id
```
I'm wondering if request_id shouldn't be required in meta data.  It seems very possible that events will be created from sources that don't originate from an HTTP request.  E.g. what if the change propagation jobs need to insert a different type of event to trigger a different type of change?

gwicke replied:
Yeah, I made the same comment a while ago, but I think it got lost. It should really be optional.

gwicke commented:
> "webrequest" is fine, just not simply "request".

I think the point is that it's not just web requests, but other kinds of requests as well. All of them _can_ have a request id to track the request processing across systems, which this field captures.

milimetric commented:
> "webrequest" is fine, just not simply "request".
> 
> I think the point is that it's not just web requests, but other kinds of
> requests as well.
> 
> Right, that's the confusion, then.  I suggest calling it "bundle_id" in
> this case.  Because "id" is already giving you a unique thing, and you're
> really just optionally sometimes grouping requests.

gwicke commented:
@milimetric: 'bundle' is not a term we use in any other context I'm aware of, so will likely be confusing.

gwicke commented on schemas/mw-page-delete.yaml:
```diff
@@ -0,0 +1,53 @@
+type: object
+properties:
+  # global event fields
+  meta:
```
Last minute alternative idea: `envelope`, as in the letter envelope. This reflects that this is primarily about sender & destination. There's even a "stamp" in the form of the event id. It's slightly clearer and more specific than `meta`, imho.

gwicke replied:
In [T116786](https://phabricator.wikimedia.org/T116786) the question of including the `topic` came up, enabling mixed-topic batch submissions. This would make this element even more of an `envelope`.

ottomata replied:
Re: `envelope`.  I prefer `meta`, but not strongly.  There is more stuff in here than just sender/receiver info, and there may be more added later (e.g. client_ip).  There also may be use cases where different types of meta data are used for different schemas (as long as some required fields always exist).  And!  I already did a bunch of code using the concept of 'meta', BUUUUT that isn't THAT hard to change ;)

gwicke replied:
> and there may be more added later (e.g. client_ip)

Sounds like the 'sender' on an envelope, no?

milimetric commented:
> @milimetric https://github.com/milimetric: 'bundle' is not a term we
> use in any other context I'm aware of, so will likely be confusing.
> 
> request_group_id ?  Last idea, I'll stop bike shedding now :)

eevans commented on schemas/mw-page-edit.yaml:
```diff
@@ -0,0 +1,70 @@
+    format: date-time
+    description: the time stamp of the revision in ISO8601 format
+  user_id:
+    type: string
```
Should `user_id` be integer here?  MediaWiki has an integer ID, a user name (Eevans), and a full name (Eric Evans).

gwicke replied:
+1 for making this an actual integer.

gwicke commented:
Two events that would be useful to add are:
- page creation
- file upload

halfak commented:
I just saw that we still have `saved_dt` instead of `timestamp` for these events.  I think that this is still problematic.  The arguments raised is that it is ambiguous, but I don't see how it is ambiguous in the context of an event.  It seems that, right now, events have a `meta` sub-object that contains it's own timestamps.  I'm not worried about how those are named since we are _inventing them_ but for the fields that we are replicating from other datastores, we should use the name that those datastores use.

So, I don't see what the problem is.  How is `mw-page-edit.timestamp` ambiguous?  What other timestamps do you plan to directly associate with the `mw-page-edit` event?  Can they go into `meta`?

halfak commented on schemas/mw-page-edit.yaml:
```diff
@@ -0,0 +1,70 @@
+  namespace:
+    type: integer
+    description: the namespace ID the page belongs to
+  revision:
```
"id", "rev_id" or "revision_id" would be more expected here.

halfak commented on schemas/mw-page-edit.yaml:
```diff
@@ -0,0 +1,70 @@
+    type: integer
+    minimum: 1
+    description: the revision ID created by this edit event
+  base_revision:
```
MediaWiki calls this "parent_id"  There's no occurrence of "base revision" that I'm aware of.

halfak commented on schemas/mw-page-edit.yaml:
```diff
@@ -0,0 +1,70 @@
+      - dt
+      - domain
+  # event-specific fields
+  title:
```
If we're including `title`, we should include `namespace`.  Also, it seems that we should give this a prefix of `page_` since the `title` is not really part of the edit event.  This is a denormalization of the concepts of _page_ and _revision_.  If `page_id`, then `page_namespace` and `page_title`.

Commit:
Some last changes based on feedback

d00rman commented:
Thank you all for your comments, I tried to incorporate most of them. I'm closing this PR and moving the schemas over to https://gerrit.wikimedia.org/r/#/c/254180/ as we have opened a new Gerrit repo for hosting schemas.
